### DISCUSSION TITLE
Excess Electricity Compensation

### DISCUSSION THREAD

**Discussion by wjohn007** [arendst/Tasmota#19095] [Show and tell]
I've been thinking about how to use excess electricity for a few months 
after realizing that my mini-photovoltaic system feeds most of it into the grid.

It all ended with a solution that can be implemented with Tasmota.

The concept:
![grafik](https://github.com/arendst/Tasmota/assets/115527085/ac6cd159-b952-4ca5-85de-533076787c2f)

The SmartMeter(1) is read out via Tasmota-Controller A (2). 
The current electrical consumption in W is available via MQTT and used by Controller B. (3)

With excess electricity, a heating cartridge for hot water (4) is to be controlled with the help of Tasmota-Controller B (Tasmota-32).
If there is no excess electricity, the water temperature must not fall below a minimum.

Using a solid-state-relais with zero-cross-detection (SSR), Controller B is used to compensate excess.

Controller B achieves this using PWM through wave packet control.

The current UI of controller B.

![grafik](https://github.com/arendst/Tasmota/assets/115527085/6471136d-2844-4b3a-8dd7-8b4ed4580c2e)

**The result:**
The electricity from photovoltaics can be almost completely utilised.

A key factor in the successful implementation was the support provided by berry scripting.

At this point a big thank you to all developers and supporters of the Tasmota project.

John

**Comment by s-hadinger** [comment]
This is awesome. Are you planning to share code and schematics?

**Reply by wjohn007** [reply]
Yes of course, once the software is hardened.
But here is the schematic for controller B
![grafik](https://github.com/arendst/Tasmota/assets/115527085/3ea62b96-c517-4d8c-b459-9464a351b5e2)

**Comment by wjohn007** [comment]
Tasmota-32 can easily be used as a power controller. 

This is possible because Tasmota-32 allows for low PWM frequencies.

With this method it is important that the mains frequency is always a multiple of the PWM frequency.

I work with a PWM frequency of 2 Hz. This allows 25 different levels to be set.

The following relationship results for the desired level

![grafik](https://github.com/arendst/Tasmota/assets/115527085/d6db01a5-1db9-40d9-a983-b0504d0da995)

**Comment by jab42** [comment]
> A key factor in the successful implementation was the support provided by berry scripting.

How about sharing the Script as it is?

**Comment by wjohn007** [comment]
The script has now become a complete application with 100 kByte of Berry source code.
Without documentation, use becomes difficult.
I plan to open a repository and store all important information there.
Have some patience.

John

**Reply by jab42** [reply]
>Have some patience.
Already did for half a Year :-)
Also without Documentation it would be a huge Source of Inspiration...

**Reply by wjohn007** [reply]
Get inspired by the first released component called  [UdpBroker](https://github.com/wjohn007/Berry-UdpBroker)   which is part of the application. More will follow.

John

**Reply by jab42** [reply]
Thanks.
I am more interested in the visualisation Part, the Tasmota Display and the computation.
So i need to wait for the Script itself...

**Reply by wjohn007** [reply]
@jab42 
The component [DallasTemp](https://github.com/wjohn007/Berry-DallasTemperature) is also used by Surgy-Box application. 
You'll find some UI stuff there.
John

**Reply by jab42** [reply]
> I plan to open a repository and store all important information there. Have some patience.
Still no News more than one and a half Year later? :)

**Comment by btsimonh** [comment]
thankyou for the useful idea.

I have an off-grid home where the max output of the invertor is 3kw, but I have a LOT of battery (~20khw).
The system cannot be licensed to feed back to the grid, so a grid connection is expensive for ONLY when we need more than 3kw peak.
So I have two problems:
1. washing machine + kettle + blender -> > 3kw -> invertor tripping out.
2. MOST of the time, the power from solar is wasted (on an average day, we may use ~4kwh, and the sun can provide ~12kwh).

Also, immersion heaters on timers with temperature sensors create an interesting power dynamic when we want to wash or dishwasher - so currently we only run one of the two, and on a timer (the one where we don't have solar water).

So I'm thinking to put power plugs on the washing machine, dishwasher.  But the SCR use on the immersion heater(s) makes so much sense....

I'll also be considering battery backup on an RPiZero and the internet connection, with MQTT retain on the power messages for the major appliances, so that if the invertor DOES trip, when it comes back up (currently complex manual procedure), the major appliances will come up to 'off'.

I suppose I could go so far as to put controlled breakers in the solar system, such that e.g. NodeRed could automatically reset the solar controller safely (I'd need a 60A relay for the DC, plus control of two breakers, and all would need to be battery backed - possible directly off the 48v (~53v) batteries via DC-DC conversion).

**Comment by wjohn007** [comment]
I'm happy if my idea helps you.
 Good luck with the project.

John

**Comment by jab42** [comment]
Still not hardened?
No Script or some Software?
Do you have a Timeline?

**Comment by wjohn007** [comment]
The solution was extensively expanded with infrared heaters. 
The same technology is used here as with the heating cartridge mentioned above.

The mini photovoltaic system was expanded with a large roof photovoltaic system.
The contribution of photovoltaic energy to heat supply is now significant.

Many Tasmota controllers are involved, so that a kind of swarm application has been created.
The controllers have to coordinate with each other when there is excess energy. 
A higher-level energy manager is not required.

All components were implemented with Tasmota, even the message broker mechanism (using UDP).

![grafik](https://github.com/user-attachments/assets/79b66be9-9e72-4255-9e26-07b2fd79e2b8)

John

**Reply by Noschvie** [reply]
Hi John, why are you not using MQTT ?

**Reply by wjohn007** [reply]
This gives me maximum independence from external systems and great resilience.
If the MQTT broker fails (as a single point of failure), nothing works anymore, neither the temperature control nor the surplus utilization.
I prefer to use MQTT for monitoring.

The message broker is a virtual device that is inherent in each controller using UDP broadcast.

John

**Reply by HofeBY** [reply]
@wjohn007

With global variables, you can transmit measured values ​​from one Tasmota device to several other
Tasmotas without the optional mqtty broker.

Even without having to configure the Tasmota measured value transmitter on its receivers.
The control loop is shifted to the receiver module.

Example see:
https://github.com/HofeBY/Tasmota-Firmware
File MT681-4xS0Impuls-Zaehler-script.txt

Another way of sending messurements by broadcast is to use KNX.

HofeBY

**Reply by wjohn007** [reply]
@HofeBY 

Thanks for the tips, but I'm really happy with my [UdpBroker](https://github.com/arendst/Tasmota/discussions/20475).
The usage is very similar to the MQTT connection in Berry. 
Since my application consists of 100% Berry code, UdpBroker is perfect for me.

With UdpBroker, the payload can be as complex or nested as you like. 
Even devices that are not part of Tasmota can participate, for example a Homematic CCU sends temperature values ​​via broadcast.
The use of scripts no longer allows rules, which I almost always use.

**Comment by NullByteZero** [comment]
Hi! I've built something very similar but the reaction time of it is not that great. Probably because total usage is measured by Shelly EM at 1Hz. At what rate is your system measuring power? Thanks in advance!

**Comment by wjohn007** [comment]
@NullByteZero 
The scan rate of the samrtmeter is 1 second. 
I prepare the value using a moving average of over 5 seconds and send it to the power controllers via UDP at 5 second intervals.
These work as 3-point controllers. 
The mechanism is simple but very stable and sufficiently efficient for me.
